skimage/feature/_canny.py

```diff
@@ -14,44 +14,175 @@
+        image is converted according to the conventions of `img_as_float`.
+        Also see https://scikit-image.org/docs/dev/user_guide/data_types.html
+
+    Returns
+    -------
+    smoothed_image : ndarray
+        The smoothed array
+    eroded_mask : ndarray
+        The eroded mask.
 
-    Notes
-    ------
-    This function calculates the fractional contribution of masked pixels
-    by applying the function to the mask (which gets you the fraction of
-    the pixel data that's due to significant points). We then mask the image
-    and apply the function. The resulting values will be lower by the
-    bleed-over fraction, so you can recalibrate by dividing by the function
-    on the mask to recover the effect of smoothing from just the significant
-    pixels.
     """
-    bleed_over = function(mask.astype(float))
-    masked_image = np.zeros(image.shape, image.dtype)
+
+    if mask is None:
+        mask = np.ones(image.shape, dtype=bool)
+
+    masked_image = np.zeros_like(image)
     masked_image[mask] = image[mask]
-    smoothed_image = function(masked_image)
-    output_image = smoothed_image / (bleed_over + np.finfo(float).eps)
-    return output_image
 
+    # Compute the fractional contribution of masked pixels by applying
+    # the function to the mask (which gets you the fraction of the
+    # pixel data that's due to significant points)
+    bleed_over = (gaussian(mask.astype(float), sigma=sigma, mode=mode)
+                  + np.finfo(float).eps)
+
+    # Smooth the masked image
+    smoothed_image = gaussian(masked_image, sigma=sigma, mode=mode)
+
+    # Lower the result by the bleed-over fraction, so you can
+    # recalibrate by dividing by the function on the mask to recover
+    # the effect of smoothing from just the significant pixels.
+    smoothed_image /= bleed_over
+
+    # Make the eroded mask. Setting the border value to zero will wipe
+    # out the image edges for us.
+    s = ndi.generate_binary_structure(2, 2)
+    eroded_mask = ndi.binary_erosion(mask, s, border_value=0)
+
+    return smoothed_image, eroded_mask
+
+
+def _set_local_maxima(magnitude, pts, w_num, w_denum, row_slices,
+                      col_slices, out):
+    # Get the magnitudes shifted left to make a matrix of the points to the
+    # right of pts. Similarly, shift left and down to get the points to the
+    # top right of pts.
+    r_0, r_1, r_2, r_3 = row_slices
+    c_0, c_1, c_2, c_3 = col_slices
+    c1 = magnitude[r_0, c_0][pts[r_1, c_1]]
+    c2 = magnitude[r_2, c_2][pts[r_3, c_3]]
+    m = magnitude[pts]
+    w = w_num[pts] / w_denum[pts]
+    c_plus = c2 * w + c1 * (1 - w) <= m
+    c1 = magnitude[r_1, c_1][pts[r_0, c_0]]
+    c2 = magnitude[r_3, c_3][pts[r_2, c_2]]
+    c_minus = c2 * w + c1 * (1 - w) <= m
+    out[pts] = c_plus & c_minus
+
+    return out
+
+
+def _get_local_maxima(isobel, jsobel, magnitude, eroded_mask):
+    #
+    # Find the normal to the edge at each point using the arctangent of the
```

Comment:
This comment seems to apply to the first block of code only, which confused me. I think a doctring would help also.